Execute plugins in a separate thread

schooft opened the pull request:
Slow, blocking or failing plugins can negatively influence data storing.

By moving all plugin code to a separate daemon thread, the main thread
is protected from misbehaving plugins.

To enable this, the data type is changed to a configuration option.

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+    try:
+        plugin_m = import_module("plugins." + plugin_name)
+        plugin = plugin_m.Plugin(plugin_config)
+        plugin.setup()
```
if there is more than a single instance of mention, it might help to define a base class; this allows for more accurate exceptions and error handling -> create a base class either for such a module, or the Plugin class

schooft replied:
Added description

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
 
+def run_plugin_thread(
+        plugin_name, plugin_config, target_dir, data_queue, log, event):
+    # Execute all plugin code on a separate thread to protect main thread from
```
it might help to give a parameter list (including expected type) of the arguments here

schooft replied:
Added docstring

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+            continue
+
+        try:
+            [metadata, data] = data
```
is the rhs object 'data' immutable? otherwise this is asking for trouble

schooft replied:
It is mutable but should not be used after processing. Added comment to clarify that.

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+    except Exception:
+        log.error(
+            "Error while stopping '%s' plugin", plugin_name, exc_info=True)
+
```
I can't see or find the threaded execution of the plugin code here!?

schooft replied:
I tried to reword the description. Is this more clear now?

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+
+        self.plugin_queue = queue.Queue()
+        self.plugin_event = threading.Event()
+        self.plugin_thread = threading.Thread(
```
ah, here the actual thread command is started

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+        self.plugin_thread.start()
+
+    def put(self, message):
+        if self.plugin_queue.qsize() < 10000:
```
magic number -> better place at top, or create alias

schooft replied:
Added comment

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+        if self.plugin_thread.is_alive():
+            self.log.error(
+                "'%s' plugin did not join within timeout", self.plugin_name)
+
```
the flow logic is not entirely clear. why does the stop() function have no return value?
right now it can be called, and it seems regardless which effect it had

schooft replied:
`stop` is called at shutdown to give the plugin thread some time to process remaining messages before it is stopped or even "killed" when the process exits. What should it return?

criticanalytic replied:
it's a kind of design question, whether the calls should return anything to at least give the opportunity to handle any result or side effect. if there are no side effects to be concerned, then it doesn't need anything

schooft replied:
Currently, I try to handle everything within the PluginHandler. Is there anything that I missed and that the external code could or should do?

criticanalytic replied:
it's OK in either way. but it should be made somewhat explicit, if any action or effect is expected of the call -- and whether something depends on it. e.g. like "fire and forget" vs "accounting"

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -280,8 +372,7 @@ def __init__(self):
-        self.plugin_config = None
-        self.plugin = None
+        self.plugin_handler = None
 
```
OK, replace the direct code with a handler

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -406,20 +497,13 @@ def _setup_logging(self):
         try:
             plugin_name = self.config["datareceiver"]["plugin"]
-            self.plugin_config = self.config[plugin_name]
+            plugin_config = self.config[plugin_name]
```
this took me a while to understand -- the plugin becomes a command line parameter?

schooft replied:
The plugin can be configured in the yaml config file for the receiver.

criticanalytic commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -406,20 +497,13 @@ def _setup_logging(self):
-            self.plugin = None
+        self.plugin_handler = PluginHandler(
+            plugin_name, plugin_config, self.target_dir, self.log)
 
```
looks good

criticanalytic commented on test/pytest/receiver/test_receiver.py:
```diff
@@ -0,0 +1,151 @@
+import pytest
+
+receiver_path = (
+    Path(__file__).parent.parent.parent.parent / "src/hidra/receiver")
```
isn't there a nicer way to write this? :-)

schooft replied:
Unfortunately, I didn't find a better easy way. A real solution is to make Hidra "pip installable" which would likely require major restructuring of the file hierarchy and the packaging scripts. This is on my list.

criticanalytic replied:
that may be not a real priority. pip is imo inferior to most other installation methods; so it might not be worth the hassle

schooft replied:
If a project is "pip installable", modules can be imported normally and IDEs will be able to navigate the code, show docstrings and autocompletions. 

I strongly recommend to always make a project "pip installable" from the beginning to avoid the problems the current Hidra setup has, e.g., the need for ugly lines like the above.

criticanalytic replied:
I'd argue that it is a byproduct. most projects are pip-installable from the beginning; but in itself is not a good requirement.
beyond that -- despite the "easy usability" and "user friendly" promises of Python I find the whole navigation and file system interaction of Python extremely irritating and annoying. I have to yet come across a project that does this in a consistent, easy-to-read way

criticanalytic commented on test/pytest/receiver/test_receiver.py:
```diff
@@ -0,0 +1,151 @@
+    plugin_handler.put((metadata, 1))
+    plugin_handler.stop()
+    mock_plugin.process.assert_called_once_with(
+        local_path="bar/relative_path/filename",
```
isn't it possible to fix this as a variable, instead of writing it manually?

schooft replied:
I could write
```
def test_plugin_put(mock_plugin):
    target_dir = "bar"
    plugin_handler = PluginHandler("mock_plugin", {"foo": 1}, target_dir, log)
    plugin_handler.start()
    plugin_handler.put((metadata, 1))
    plugin_handler.stop()
    local_path = os.path.join(target_dir, metadata["relative_path"], metadata["filename"])
    mock_plugin.process.assert_called_once_with(
        local_path=local_path,
        metadata=metadata,
        data=1)
```
but I think this would make the test less readable without a clear advantage.

criticanalytic replied:
sorry, that was misleading. I only meant the single line of assignment test of 'local_path'

criticanalytic commented on test/pytest/receiver/test_receiver.py:
```diff
@@ -0,0 +1,151 @@
+    plugin_handler.stop(0)
+    assert time.time() - t0 < 10
+    assert plugin_handler.plugin_queue.qsize() == 10000
+    assert "Skipping" in caplog.text
```
are the assertion tests of the time to check whether the plugin delay in a thread causes a delay outside it?

schooft replied:
Partly. When using a queue with an internal maximum size, `put` blocks when the queue is full. This tests asserts that the manual queue size limit is non-blocking. I added a docstring to clarify.

criticanalytic reviewed (approved):
looks ok, only comments:
- only one piece of code potentially "dangerous" if assigned variable is mutable
- add parameter meanings to argument list of module global function to make reading easier
- think about making an abstract plugin for easier handling
- design unclear if the handler calls should or should not return something

schooft commented:
Thanks for the review!

> only one piece of code potentially "dangerous" if assigned variable is mutable

Good catch! Currently, the mutable variable is not reused and I added a comment to clarify that this would be dangerous

>  add parameter meanings to argument list of module global function to make reading easier

Done

> think about making an abstract plugin for easier handling

I agree that the plugin implementation is incomplete. However, there is currently only one plugin and I don't have plans to add more. Duck typing should be sufficient for now. I added a description of the informal interface.

> design unclear if the handler calls should or should not return something

The only method with a non-None return value is `get_data_type`. Which other calls could/should return something?

schooft commented on src/hidra/receiver/datareceiver.py:
```diff
@@ -259,6 +260,97 @@ def stop(self):
+    # Execute all plugin code on a separate thread to protect main thread from
+    # slow or blocking plugins
+    try:
+        plugin_m = import_module("plugins." + plugin_name)
```
Added description

criticanalytic reviewed (approved):
all clear now
